File: Java/SkipList/SkipList.java
package SkipList;
/*
 * @author Ionésio Junior
 * */
/*
 * Skip List Implementation
 * */
public class SkipList<T extends Comparable<T>> implements SkipListInterface<T> {
	private SkipListNode<T> root;
	private SkipListNode<T> NIL;
	private int height;
	
	public SkipList(int maxHeight){
		this.height = maxHeight;
		this.root   = new SkipListNode<T>(Integer.MIN_VALUE, maxHeight, null);
		this.NIL = new SkipListNode<T>(Integer.MAX_VALUE, maxHeight,null);
		connectRootToNil();
	}
	
	private void connectRootToNil(){
		for(int i = 0 ; i < root.getForward().length;i++){
			root.getForward()[i] = NIL;
		}
	}

	@Override
	public void insert(int key, T newValue, int height) {
		if(newValue != null){
			this.ajustHeight(height);
			SkipListNode<T>[] previousNodes = new SkipListNode[height];
			SkipListNode<T> aux = this.root;
			for(int i = height - 1 ; i >= 0;i--){
				while(aux.getForward(i).getValue() != null && aux.getForward(i).getKey() < key){
					aux = aux.getForward(i);
				}
				previousNodes[i] = aux;
			}
			aux = aux.getForward(0);
			if(aux.getKey() == key){
				aux.setValue(newValue);
			}else{
				aux = new SkipListNode<T>(key, height, newValue);
				changePointers(height,previousNodes,aux);
			}
		}
	}

     @Override
     public void remove(int key) {
        SkipListNode[] array = new SkipListNode[this.height];
        SkipListNode<T> aux = this.root;
        for (int i = this.height - 1; i >= 0; i--) {
           if (aux.getForward(i) != this.NIL) {
              while (aux.getForward(i).getValue() != null && aux.getForward(i).getKey() < key)
                 aux = aux.getForward(i);
           }
           array[i] = aux;
        }
        aux = aux.getForward()[0];
  
        if (aux.getKey() == key) {
  
           for (int i = 0; i < this.height; i++) {
              if (array[i].getForward()[i] != aux) {
                 break;
              }
             array[i].getForward()[i] = aux.getForward()[i];
          }
       }
    }

	@Override
	public int height() {
		int height = this.height - 1;
		while(height >= 0 && this.root.getForward(height) == NIL){
			height--;
		}
		return height+ 1;
	}

	@Override
	public SkipListNode<T> search(int key) {
		SkipListNode<T> aux = this.root;
		for(int i = this.height - 1;i >= 0;i--){
			while(aux.getForward(i).getValue() != null && aux.getForward(i).getKey() < key){
				aux = aux.getForward(i);
			}
		}
		SkipListNode<T> result = aux.getForward(0);
		if(result.getKey() != key){
			return null;
		}else{
			return result;
		}
	}

	@Override
	public int size() {
		int count = 0; 
		SkipListNode<T> aux  = this.root.getForward(0);
		while(aux != NIL){
			count++;
			aux = aux.getForward(0);
		}
		return count;
	}

	@Override
	public SkipListNode<T>[] toArray() {
		int size = this.size() + 2;
		SkipListNode<T>[] result = new SkipListNode[size];
		SkipListNode<T> aux = this.root;
		int index = 0;
		 while(index < size){
			 result[index++] = aux;
			 aux = aux.getForward(0);
		 }
		 return result;
	}
	
	///////////////////////   AUXILIAR METHODS		//////////////////////////////
	
	private void ajustHeight(int height){
		if(height > this.height){
			SkipListNode<T>[] rootArray = new SkipListNode[height];
			this.NIL.setForward(new SkipListNode[height]);
			for(int i = 0; i < height;i++){
				if(i < this.height){
					rootArray[i] = this.root.getForward(i);
				}else{
					rootArray[i] = this.NIL;
				}
			}
			this.root.setForward(rootArray);
			this.height = height;
		}
	}
	
	
	private void changePointers(int height,SkipListNode[] previousNodes,SkipListNode<T> next){
		for(int i = 0; i < height;i++){
			if(previousNodes[i].getForward(i) == null){
				next.getForward()[i] = NIL;
			}else{
				next.setForward(i, previousNodes[i].getForward(i));
				previousNodes[i].setForward(i, next);
			}
		}
	}
	
}

File: Python/SkipList/SkipList.py
'''
	Author: Ionesio Junior
'''
import sys
from Node import Node


'''
	SkipList  class  implementation
'''
class SkipList(object):
	
	def __init__(self,maxHeight):
		self.__height = maxHeight
		self.__head = Node(-1 * sys.maxint,maxHeight,None)
		self.__tail = Node(sys.maxint,maxHeight,None)
		self.__connectHeadToTail()

	def __connectHeadToTail(self):
		for i in range(self.__height):
			self.__head.setForwardNode(i,self.__tail)
	
	
	def insert(self,key,data,height):
		if(data != None and key != None and height != None):
			self.__ajustHeight(height)
			previousNodes = [None] * height
			aux = self.__head
			for i in range(height - 1,-1,-1):
				while(aux.getForwardNode(i).getValue() != None and aux.getForwardNode(i).getKey() < key):
					aux = aux.getForwardNode(i)
				
				previousNodes[i] = aux
			aux = aux.getForwardNode(0)
			if(aux.getKey() == key):
				aux.setData(data)
			else:
				aux = Node(key,height,data)
				self.__changePointers(height,previousNodes,aux)
	
	def __changePointers(self,height,previousNodes,aux):
		for i in range(height):
			if(previousNodes[i].getForwardNode(i) == None):
				aux.setForwardNode(i,self.__tail)
			else:
				aux.setForwardNode(i,previousNodes[i].getForwardNode(i))
				previousNodes[i].setForwardNode(i,aux)


	def __ajustHeight(self,height):
		if(height > self.__height):
			self.__tail.setForward([None] * height)
			newForward = [None] * height
			for i in range(height):
				if(i < self.__height):
					newForward[i] = self.__head.getForwardNode(i)
				else:
					newForward[i] = self.__tail
			self.__height = height
			self.__head.setForward(newForward)
	
	
	def remove(self,key):
		array  = [None] * self.__height
		aux = self.__head
		for i in range(self.__height - 1, -1,-1):
			while(aux.getForwardNode(i).getValue() != None and aux.getForwardNode(i).getKey() < key):
				aux = aux.getForwardNode(i)
			array[i] = aux
		aux = aux.getForwardNode(0)
		if(aux.getKey()  == key):
			for i in range(self.__height):
				if(array[i].getForwardNode(i).getKey() != aux.getKey()):
						break;
				array[i].setForwardNode(i,aux.getForwardNode(i))

	def search(self,key):
		aux = self.__head
		for i in range(self.__height - 1,-1,-1):
			while(aux.getForwardNode(i).getValue() != None and aux.getForwardNode(i).getKey() < key):
				aux = aux.getForwardNode(i)
		aux = aux.getForwardNode(0)
		if(aux.getKey() == key):
			return aux
		else:
			return None

	def size(self):
		count = 0
		aux = self.__head.getForwardNode(0)
		while(aux.getKey() != self.__tail.getKey()):
			count += 1
			aux = aux.getForwardNode(0)
		return count

	def toList(self):
		size = self.size() + 2
		nodeList = []
		index = 0
		aux = self.__head
		while(index < size):
			nodeList.append((aux.getValue(),aux.getKey(),aux.height()))
			aux = aux.getForwardNode(0)
			index += 1
		return nodeList
